src/CPUActivityAgent.cpp

```diff
@@ -0,0 +1,564 @@
+                m_do_write_batch = true;
+            }
+        }
+
+        // Set per package controls
+        for (int domain_idx = 0; domain_idx < M_NUM_PACKAGE; ++domain_idx) {
+            if(std::isnan(uncore_freq_request.at(domain_idx))) {
+                uncore_freq_request.at(domain_idx) = in_policy[M_POLICY_UNCORE_FREQ_MAX];
+            }
+
+            if (uncore_freq_request.at(domain_idx) !=
+                m_uncore_freq_min_control.at(domain_idx).last_setting ||
+                uncore_freq_request.at(domain_idx) !=
+                m_uncore_freq_max_control.at(domain_idx).last_setting) {
+                // Adjust
+                m_platform_io.adjust(m_uncore_freq_min_control.at(domain_idx).batch_idx,
+                                    uncore_freq_request.at(domain_idx));
+
+                m_platform_io.adjust(m_uncore_freq_max_control.at(domain_idx).batch_idx,
+                                    uncore_freq_request.at(domain_idx));
+
+                // Save the value for future comparison
+                m_uncore_freq_min_control.at(domain_idx).last_setting = uncore_freq_request.at(domain_idx);
+                m_uncore_freq_max_control.at(domain_idx).last_setting = uncore_freq_request.at(domain_idx);
+                ++m_uncore_frequency_requests;
+
+                m_do_write_batch = true;
+            }
+        }
+    }
+
+    // If controls have a valid updated value write them.
+    bool CPUActivityAgent::do_write_batch(void) const
+    {
+        return m_do_write_batch;
+    }
+
+    // Read signals from the platform and calculate samples to be sent up
+    void CPUActivityAgent::sample_platform(std::vector<double> &out_sample)
+    {
+        GEOPM_DEBUG_ASSERT(out_sample.size() == M_NUM_SAMPLE,
+                           "CPUActivityAgent::" + std::string(__func__) +
+                           "(): sample vector not correctly sized.  Expected  " +
+                            std::to_string(M_NUM_SAMPLE) + ", actual: " +
+                            std::to_string(out_sample.size()));
+
+        // Collect latest signal values
+        for (int domain_idx = 0; domain_idx < M_NUM_PACKAGE; ++domain_idx) {
+            // Frequency signals
+            m_uncore_freq_status.at(domain_idx).value = m_platform_io.sample(m_uncore_freq_status.at(domain_idx).batch_idx);
+
+            // Uncore steering signals
+            m_qm_rate.at(domain_idx).value = m_platform_io.sample(m_qm_rate.at(domain_idx).batch_idx);
+        }
+
+        for (int domain_idx = 0; domain_idx < M_NUM_CORE; ++domain_idx) {
+            // Core steering signals
+            m_core_scal.at(domain_idx).value = m_platform_io.sample(m_core_scal.at(domain_idx).batch_idx);
+        }
+    }
+
+    // Wait for the remaining cycle time to keep Controller loop cadence
+    void CPUActivityAgent::wait(void)
+    {
+        geopm_time_s current_time;
+        do {
+            geopm_time(&current_time);
+        }
```

Comment:
Do we really need a busy wait? It seems like this might be done to reduce variation in time between controller steps. Does this agent actually need that? All of the sampled values are read as rates anyways (vs point-in-time snapshots of monotonic counters).

Reply:
Yeah, this is just the standard wait implementation all agents are using. We've talked about standardizing the implementation and factoring it out so that it doesn't have to be implemented in every agent. We've talked about a smarter implementation that leverages nanosleep() to minimize the amount of time we rely on a busy wait.

I would prefer to defer addressing this to when this refactoring work is done, as it will address it for all the agents.

Reply:
Sounds good to defer this comment to #1065